IPC stream send is slow due to semaphore signal if receiver is faster than sender

kkinnunen-apple opened the pull request:
#### 05cb01bde6d067ee099500456f25a378abec4fd1
<pre>
IPC stream send is slow due to semaphore signal if receiver is faster than sender
<a href="https://bugs.webkit.org/show_bug.cgi?id=239895">https://bugs.webkit.org/show_bug.cgi?id=239895</a>

Reviewed by Wenson Hsieh.

Add a message property `StreamBatched`. Consecutive messages marked as such
will not wake up the sleeping receiver. This skips the semaphore signal call
in cases that the receiver is faster than the sender.

Conservatively add batched property only to cases that are invoked by MotionMark
Canvas test case.
Conservatively use maximum batch size of 40, this seems to skip most signal
invocations.

* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/Platform/IPC/StreamClientConnection.cpp:
(IPC::StreamClientConnection::setSemaphores):
(IPC::StreamClientConnection::wakeUpServer):
(IPC::StreamClientConnection::wakeUpServerBatched):
(IPC::StreamClientConnection::deferredWakeUpServer): Deleted.
(IPC::StreamClientConnection::decrementRemainingMessageCountBeforeSendingWakeUp): Deleted.
(IPC::StreamClientConnection::sendDeferredWakeUpMessageIfNeeded): Deleted.
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendStream):
(IPC::StreamClientConnection::trySendSyncStream):
(IPC::StreamClientConnection::trySendDestinationIDIfNeeded):
(IPC::StreamClientConnection::sendProcessOutOfStreamMessage):
* Source/WebKit/Scripts/webkit/messages.py:
(message_to_struct_declaration):
(generate_header_includes_from_conditions):
* Source/WebKit/Scripts/webkit/tests/Makefile:
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::jsValueForArguments):
(IPC::messageArgumentDescriptions):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
(IPC::description):
(IPC::receiverName):
(IPC::isValidMessageName):
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithIfMessageMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithIfReceiver.messages.in: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithIfReceiverMessageReceiver.cpp: Copied from Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp.
(WebKit::TestWithIfReceiver::didReceiveMessage):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithIfReceiver_CompletePaymentSession&gt;):
* Source/WebKit/Scripts/webkit/tests/TestWithIfReceiverMessages.h: Copied from Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp.
(Messages::TestWithIfReceiver::messageReceiverName):
(Messages::TestWithIfReceiver::CompletePaymentSession::name):
(Messages::TestWithIfReceiver::CompletePaymentSession::CompletePaymentSession):
(Messages::TestWithIfReceiver::CompletePaymentSession::arguments const):
* Source/WebKit/Scripts/webkit/tests/TestWithIfReceiverMessagesReplies.h: Copied from Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp.
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatched.messages.in: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatchedMessageReceiver.cpp: Copied from Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp.
(WebKit::TestWithStreamBatched::didReceiveStreamMessage):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithStreamBatched_SendString&gt;):
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatchedMessages.h: Copied from Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp.
(Messages::TestWithStreamBatched::messageReceiverName):
(Messages::TestWithStreamBatched::SendString::name):
(Messages::TestWithStreamBatched::SendString::SendString):
(Messages::TestWithStreamBatched::SendString::arguments const):
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatchedMessagesReplies.h: Copied from Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp.
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessageReceiver.cpp:

Canonical link: <a href="https://commits.webkit.org/251815@main">https://commits.webkit.org/251815@main</a>
</pre>

whsieh commented on Source/WebKit/Platform/IPC/StreamClientConnection.h:
```diff
@@ -160,8 +157,8 @@ class StreamClientConnection final {
-    unsigned m_remainingMessageCountBeforeSendingWakeUp { 0 };
-    unsigned m_wakeUpMessageHysteresis { 0 };
+    unsigned m_maxBatchSize = 20; // Number of messages marked as StreamBatched to accumulate before notifying the server.
+    unsigned m_batchSize = 0;
```
Nit - I think we generally prefer `{ }` for initializers, rather than `=`.

kkinnunen-apple committed:
IPC stream send is slow due to semaphore signal if receiver is faster than sender
https://bugs.webkit.org/show_bug.cgi?id=239895

Reviewed by Wenson Hsieh.

Add a message property `StreamBatched`. Consecutive messages marked as such
will not wake up the sleeping receiver. This skips the semaphore signal call
in cases that the receiver is faster than the sender.

Conservatively add batched property only to cases that are invoked by MotionMark
Canvas test case.
Conservatively use maximum batch size of 40, this seems to skip most signal
invocations.

* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/Platform/IPC/StreamClientConnection.cpp:
(IPC::StreamClientConnection::setSemaphores):
(IPC::StreamClientConnection::wakeUpServer):
(IPC::StreamClientConnection::wakeUpServerBatched):
(IPC::StreamClientConnection::deferredWakeUpServer): Deleted.
(IPC::StreamClientConnection::decrementRemainingMessageCountBeforeSendingWakeUp): Deleted.
(IPC::StreamClientConnection::sendDeferredWakeUpMessageIfNeeded): Deleted.
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendStream):
(IPC::StreamClientConnection::trySendSyncStream):
(IPC::StreamClientConnection::trySendDestinationIDIfNeeded):
(IPC::StreamClientConnection::sendProcessOutOfStreamMessage):
* Source/WebKit/Scripts/webkit/messages.py:
(message_to_struct_declaration):
(generate_header_includes_from_conditions):
* Source/WebKit/Scripts/webkit/tests/Makefile:
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::jsValueForArguments):
(IPC::messageArgumentDescriptions):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
(IPC::description):
(IPC::receiverName):
(IPC::isValidMessageName):
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithIfMessageMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithIfReceiver.messages.in: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithIfReceiverMessageReceiver.cpp: Copied from Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp.
(WebKit::TestWithIfReceiver::didReceiveMessage):
(IPC::jsValueForDecodedMessage<MessageName::TestWithIfReceiver_CompletePaymentSession>):
* Source/WebKit/Scripts/webkit/tests/TestWithIfReceiverMessages.h: Copied from Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp.
(Messages::TestWithIfReceiver::messageReceiverName):
(Messages::TestWithIfReceiver::CompletePaymentSession::name):
(Messages::TestWithIfReceiver::CompletePaymentSession::CompletePaymentSession):
(Messages::TestWithIfReceiver::CompletePaymentSession::arguments const):
* Source/WebKit/Scripts/webkit/tests/TestWithIfReceiverMessagesReplies.h: Copied from Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp.
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatched.messages.in: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatchedMessageReceiver.cpp: Copied from Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp.
(WebKit::TestWithStreamBatched::didReceiveStreamMessage):
(IPC::jsValueForDecodedMessage<MessageName::TestWithStreamBatched_SendString>):
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatchedMessages.h: Copied from Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp.
(Messages::TestWithStreamBatched::messageReceiverName):
(Messages::TestWithStreamBatched::SendString::name):
(Messages::TestWithStreamBatched::SendString::SendString):
(Messages::TestWithStreamBatched::SendString::arguments const):
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBatchedMessagesReplies.h: Copied from Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp.
* Source/WebKit/Scripts/webkit/tests/TestWithStreamBufferMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessageReceiver.cpp:

Canonical link: https://commits.webkit.org/251815@main
